test/utils/circuit.js

```diff
@@ -0,0 +1,133 @@
+'use strict'
+
+const chai = require('chai')
+const dirtyChai = require('dirty-chai')
+const expect = chai.expect
+chai.use(dirtyChai)
+
+const waterfall = require('async/waterfall')
+const crypto = require('crypto')
+
+const IPFS = require('ipfs')
+
+const DaemonFactory = require('ipfsd-ctl')
+const jsDf = DaemonFactory.create({ type: 'js' })
+const goDf = DaemonFactory.create({ type: 'go' })
+const procDf = DaemonFactory.create({ type: 'proc', exec: IPFS })
+
+const baseConf = {
+  Bootstrap: [],
+  Addresses: {
+    API: '/ip4/0.0.0.0/tcp/0',
+    Gateway: '/ip4/0.0.0.0/tcp/0'
+  },
+  Discovery: {
+    MDNS: {
+      Enabled:
+        false
+    }
+  }
+}
+
+exports.setUpProcNode = (addrs, callback) => {
+  procDf.spawn({
+    initOptions: { bits: 512 },
+    config: Object.assign({}, baseConf, {
+      Addresses: {
+        Swarm: addrs
+      },
+      EXPERIMENTAL: {
+        relay: {
+          enabled: true,
+          hop: {
+            enabled: true
+          }
+        }
+      }
+    })
+  }, (err, ipfsd) => {
+    expect(err).to.not.exist()
+    ipfsd.api.id((err, id) => {
+      callback(err, { ipfsd, addrs: id.addresses })
+    })
+  })
+}
+
+exports.setUpJsNode = (addrs, callback) => {
+  jsDf.spawn({
+    initOptions: { bits: 512 },
+    config: Object.assign({}, baseConf, {
+      Addresses: {
+        Swarm: addrs
+      },
+      EXPERIMENTAL: {
+        relay: {
+          enabled: true,
+          hop: {
+            enabled: true
+          }
+        }
+      }
+    })
+  }, (err, ipfsd) => {
+    expect(err).to.not.exist()
+    ipfsd.api.id((err, id) => {
+      callback(err, { ipfsd, addrs: id.addresses })
+    })
+  })
+}
+
+exports.setUpGoNode = (addrs, callback) => {
+  goDf.spawn({
+    initOptions: { bits: 1024 },
+    config: Object.assign({}, baseConf, {
+      Addresses: {
+        Swarm: addrs
+      },
+      Swarm: {
+        DisableRelay: false,
+        EnableRelayHop: true
+      }
+    })
+  }, (err, ipfsd) => {
+    expect(err).to.not.exist()
+    ipfsd.api.id((err, id) => {
+      const addrs = [].concat(id.addresses, [`/p2p-circuit/ipfs/${id.id}`])
+      callback(err, { ipfsd, addrs: addrs })
+    })
+  })
+}
+
+const data = crypto.randomBytes(128)
+exports.send = (nodeA, nodeB, callback) => {
+  waterfall([
+    (cb) => nodeA.files.add(data, cb),
+    (res, cb) => nodeB.files.cat(res[0].hash, cb),
+    (buffer, cb) => {
+      expect(buffer).to.deep.equal(data)
+      cb()
+    }
+  ], callback)
+}
+
+exports.wsAddr = (addrs) => addrs
```

Comment:
perhaps prefix these calls with `get`, so that it is `getWSAddr`. I wasn't sure what they did while reading the test file.